aspnetcore/blazor/fundamentals/environments.md

```diff
@@ -135,57 +152,208 @@ In the following example for IIS, the custom header (`Blazor-Environment`) is ad
 
-:::code language="razor" source="~/../blazor-samples/3.1/BlazorSample_WebAssembly/Pages/environments/ReadEnvironment.razor" highlight="3,7":::
+* Visual Studio: Right-click each project and select **Add** > **Project Reference**. Select the checkbox for the shared project and select **OK**.
+* Visual Studio Code/.NET CLI: Execute `dotnet add reference {PATH}` in a command shell from each of the server and `.Client` project folders. The `{PATH}` placeholder is the path to the shared class library project.
+
+In the shared (`.Shared`) project:
+
+* Add an interface for the service abstraction.
+* Add classes for the service implementation and host environment extensions.
+
+`BlazorHostEnvironment.cs`:
+
+```csharp
+public interface IBlazorHostEnvironment
+{
+    string EnvironmentName { get; set; }
+}
+
+public class BlazorHostEnvironment(string environment) : IBlazorHostEnvironment, 
+    IDisposable
+{
+    private bool _disposed;
+
+    public string EnvironmentName { get; set; } = environment;
+
+    public void Dispose()
+    {
+        if (_disposed)
+        {
+            return;
+        }
+
+        _disposed = true;
+    }
+}
+
+public static class BlazorHostEnvironmentExtensions
+{
+    public static bool IsDevelopment(this IBlazorHostEnvironment hostEnvironment)
+    {
+        return hostEnvironment.IsEnvironment("Development");
+    }
+
+    public static bool IsStaging(this IBlazorHostEnvironment hostEnvironment)
+    {
+        return hostEnvironment.IsEnvironment("Staging");
+    }
+
+    public static bool IsProduction(this IBlazorHostEnvironment hostEnvironment)
+    {
+        return hostEnvironment.IsEnvironment("Production");
+    }
+
+    public static bool IsEnvironment(
+        this IBlazorHostEnvironment hostEnvironment,
+        string environmentName)
+    {
+        return string.Equals(
+            hostEnvironment.EnvironmentName,
+            environmentName,
+            StringComparison.OrdinalIgnoreCase);
+    }
+}
+```
+
+In the server project's `Program` file, register a `IBlazorHostEnvironment` service with the `BlazorHostEnvironment` implementation. Set the host environment with <xref:Microsoft.AspNetCore.Builder.WebApplicationBuilder.Environment%2A?displayProperty=nameWithType>, which is only available on the server:
+
+```csharp
+builder.Services.AddSingleton<IBlazorHostEnvironment>(
+    sp => new BlazorHostEnvironment(builder.Environment.EnvironmentName));
+```
+
+In the `.Client` project's `Program` file, register a `IBlazorHostEnvironment` service with the `BlazorHostEnvironment` implementation. Set the host environment with <xref:Microsoft.AspNetCore.Components.WebAssembly.Hosting.IWebAssemblyHostEnvironment.Environment%2A?displayProperty=nameWithType>, which is only available on the client:
+
+```csharp
+builder.Services.AddSingleton<IBlazorHostEnvironment>(
+    sp => new BlazorHostEnvironment(builder.HostEnvironment.Environment));
```

Comment:
This is definitely valid, albeit different than what I had in mind.

If the customer were to add additional members to the `IBlazorHostEnvironment` interface, it would require adding more parameters to the primary constructor and providing values in `Program.cs`. Or if the interface had any methods in it, it would require falling back to separate implementations (unless you do a thing where you dynamically check which platform you're on using `OperatingSystem.IsBrowser()`).

I think it would be a bit more conventional to have separate implementations of `IBlazorHostEnvironment` that can inject app-specific dependencies and use those to support aspects of the implementation. For example:

`.Client` project:
```csharp
// IBlazorHostEnvironment.cs
public interface IBlazorHostEnvironment
{
    string EnvironmentName { get; }
}

// BlazorWebAssemblyHostEnvironment.cs
public class BlazorWebAssemblyHostEnvironment(IWebAssemblyHostEnvironment hostEnvironment) : IBlazorHostEnvironment
{
    public string EnvironmentName => hostEnvironment.Environment;
}

// Program.cs
builder.Services.AddSingleton<IBlazorHostEnvironment, BlazorWebAssemblyHostEnvironment>();
```

Server project:
```csharp
// BlazorServerHostEnvironment.cs
public class BlazorServerHostEnvironment(IWebHostEnvironment hostEnvironment) : IBlazorHostEnvironment
{
    public string EnvironmentName => hostEnvironment.EnvironmentName;
}

// Program.cs
builder.Services.AddScoped<IBlazorHostEnvironment, BlazorServerHostEnvironment>();
```

What do you think?

Reply:
I did it that way because I felt there probably wouldn't be anything else, but I think the most general approach is what you're suggesting. Since we'll use this as a promonient example of how to address this problem, I think your idea is best.

WRT the dispose situation from above. I'm confused if the DI container is creating the `new`-ed up services like this. I thought that it was NOT creating them ... that the dev code is creating them. If that's right, then the disposal code is boilerplate code that doesn't dispose of anything but perhaps present as a good pattern so as to be clear that the service implementation needs to dispose of anything it creates ... just as a reminder ... a pattern. Alternatively if I'm incorrect about this service registration and what's actually happening is that the DI container is creating them, then yeah ... for sure there's no need to place this boilerplate do-nothing disposal code ... such as what we'll be changing over to when I adopt your suggestion here for the serivces.

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-8.0#disposal-of-services